### PR TITLE
Add class for components

### PR DESCRIPTION
**Pull request by LuomaJuha**
Adds a class for components, so adjusting styles locally is easier. Small adjustments to the code.

### CODE REVIEW

File: themes/root/theme.config.php
Comment (demiankatz) [inline themes/root/theme.config.php]: I'm not opposed to changing from strings to class references, but it feels a little weird to do it in this one line and not convert the rest. Maybe it would be best to stick to the existing convention for this PR, and then open a separate one to convert everything (as I think we may have discussed in the past).

File: module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/ComponentTest.php
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/ComponentTest.php]: You can simplify to:

```suggestion
        $themeInfoMock = $this->createMock(ThemeInfo::class);
```

File: themes/bootstrap5/templates/_ui/components/hide-offcanvas-button.phtml
Comment (demiankatz) [inline themes/bootstrap5/templates/_ui/components/hide-offcanvas-button.phtml]: I notice that `data-toggle` is changing to `data-bs-toggle` here. Is this intentional, or is the automatic bootstrap3/bootstrap5 conversion doing something unexpected? (Just asking because the templates are now inconsistent across the two themes, but they previously were not, and yet the automated checks are passing).

File: themes/bootstrap5/templates/_ui/components/show-sidebar-button.phtml
Comment (demiankatz) [inline themes/bootstrap5/templates/_ui/components/show-sidebar-button.phtml]: Any particular reason you converted the other templates to use the `htmlAttributes` helper but left this alone? It's fine as-is, but conversion might improve readability here as well.

File: module/VuFind/src/VuFind/View/Helper/Root/ComponentFactory.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/ComponentFactory.php]: This is a lot of author credits for a simple factory. You probably can take sole credit here if you want to.

File: module/VuFind/src/VuFind/View/Helper/Root/Component.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/Component.php]: Based on the above, it looks like `$name` may contain slash characters; do those need to be normalized to make this a valid CSS class?

Also, where does the 'vc-' prefix come from? Does that mean VuFind® component? I don't object to the prefix, but just want to be sure I understand it. And maybe a comment would be in order to explain.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @LuomaJuha -- see below for a few questions, comments and suggestions.

File: themes/bootstrap5/templates/_ui/components/hide-offcanvas-button.phtml
Comment (LuomaJuha) [reply themes/bootstrap5/templates/_ui/components/hide-offcanvas-button.phtml]: hmm, good catch as i used to update-bootstrap5 command with these, ill check

File: themes/bootstrap5/templates/_ui/components/show-sidebar-button.phtml
Comment (LuomaJuha) [reply themes/bootstrap5/templates/_ui/components/show-sidebar-button.phtml]: hmm, was between is there too much adjustments in this one file for this pr. Can adjust this also

File: themes/bootstrap5/templates/_ui/components/hide-offcanvas-button.phtml
Comment (LuomaJuha) [reply themes/bootstrap5/templates/_ui/components/hide-offcanvas-button.phtml]: seems like it was data-toggle="collapse" in 3.4.1 and data-bs-toggle="collapse"  in 5. I think the update script thought that this data was bootstrap specific. Will adjust

Comment (demiankatz) [reply themes/bootstrap5/templates/_ui/components/hide-offcanvas-button.phtml]: Sounds good. And if the update script won't let you for some reason, you might want to talk to @EreMaijala about it.

Comment (EreMaijala) [reply themes/bootstrap5/templates/_ui/components/hide-offcanvas-button.phtml]: Ah, yes, there's a conflict in bootstrap3 where data-toggle is used by both VuFind and Bootstrap. Fortunately it won't be a problem that much longer. :)

File: module/VuFind/src/VuFind/View/Helper/Root/Component.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/Component.php]: One more thought: might there be an advantage to putting a `vufind-theme-x` class on the `<body>` tag in the layout, and then leaving the theme name out of the component class? That way, you could target styles for the component in general, or have more specific theme-focused styles. It would make individual class names shorter and less redundant. Or is there a reason it's better to do it this way?

Also, for $normalizedName, might it be better to do a preg_replace so you can replace any characters that are not legal in a class name? e.g. something like `[^-_A-Za-z0-9]`?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for addressing all of my comments, @LuomaJuha -- just a couple more thoughts now...

File: module/VuFind/src/VuFind/View/Helper/Root/Component.php
Comment (LuomaJuha) [reply module/VuFind/src/VuFind/View/Helper/Root/Component.php]: hmm, that could actually be quite nice. Will adjust!

**Review by EreMaijala — CHANGES_REQUESTED** [review]
Now that I gave this another thought I'd like a bit of explanation why this is needed and a good thing. I'm not sure the possible benefits outweigh the potential confusion this can cause if a theme's styles target another.

**Comment by LuomaJuha** [comment]
Reverted the changes for theme-specificity, as it could have caused confusion, but left the _componentClass to be used later for overwriting styles

File: module/VuFind/src/VuFind/View/Helper/Root/Component.php
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/Component.php]: A little more specificity here might be helpful:

```suggestion
    /**
     * Local cache of CSS class names for components (component name => class name).
     *
     * @var array
     */
    protected array $cssClassCache = [];
```

Comment (demiankatz) [reply module/VuFind/src/VuFind/View/Helper/Root/Component.php]: ...and obviously if you agree to change the name of the property, you'll have to adjust other references accordingly. :-)

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for the simplification, @LuomaJuha! Just one last documentation-oriented suggestion below.

File: module/VuFind/src/VuFind/View/Helper/Root/Component.php
Comment (LuomaJuha) [reply module/VuFind/src/VuFind/View/Helper/Root/Component.php]: sure, old habit naming with shorter names, descriptive names are nice.

**Review by demiankatz — APPROVED** [review]
Thanks, @LuomaJuha -- looks good to me, and all tests are still passing.

I can't merge this without dismissing @EreMaijala's open review, so I'll wait and see if he has any final thoughts before finalizing the work.

**Comment by LuomaJuha** [comment]
Sure!

**Review by EreMaijala — APPROVED** [review]

**Pull request merged**
